internal/mcpserver/agent-pr-server/mcpserver.go

```diff
@@ -0,0 +1,279 @@
+	creds  *PluralCredentials
+}
+
+// NewMCPServer creates a new MCP server instance
+func NewMCPServer(creds *PluralCredentials) *MCPServer {
+	// Create a new MCP server
+	s := server.NewMCPServer(
+		"Plural Console GraphQL MCP Server",
+		"1.0.0",
+		server.WithToolCapabilities(true),
+	)
+
+	mcpServer := &MCPServer{
+		server: s,
+		creds:  creds,
+	}
+
+	// Register tools
+	mcpServer.registerTools()
+
+	return mcpServer
+}
+
+// LoadPluralCredentials loads and validates Plural credentials from environment variables
+func LoadPluralCredentials() (*PluralCredentials, error) {
+	accessToken := os.Getenv(controller.EnvConsoleToken)
+	if accessToken == "" {
+		return nil, fmt.Errorf("%s environment variable is required", controller.EnvConsoleToken)
+	}
+
+	consoleURL := os.Getenv(controller.EnvConsoleURL)
+	if consoleURL == "" {
+		return nil, fmt.Errorf("%s environment variable is required", controller.EnvConsoleURL)
+	}
+
+	// Ensure the console URL has the correct GraphQL endpoint
+	if !strings.HasSuffix(consoleURL, "/gql") && !strings.HasSuffix(consoleURL, "/ext/gql") {
+		// Add the GraphQL endpoint if not present
+		if strings.HasSuffix(consoleURL, "/") {
+			consoleURL += "ext/gql"
+		} else {
+			consoleURL += "/ext/gql"
+		}
+	}
+
+	return &PluralCredentials{
+		AccessToken: accessToken,
+		ConsoleURL:  consoleURL,
+	}, nil
+}
+
+// Start starts the MCP server with stdio transport
+func (m *MCPServer) Start() error {
+	klog.InfoS("Started Plural Console MCP Server", "consoleURL", m.creds.ConsoleURL)
+	return server.ServeStdio(m.server)
+}
+
+// registerTools registers all available tools with the MCP server
+func (m *MCPServer) registerTools() {
+	// Define the agentPullRequest tool
+	tool := mcp.NewTool("agentPullRequest",
+		mcp.WithDescription("Create a pull request through the Plural console GraphQL API for agent-generated changes"),
+		mcp.WithString("runId",
+			mcp.Required(),
+			mcp.Description("The agent run ID associated with this pull request"),
+		),
+		mcp.WithString("repository",
+			mcp.Required(),
+			mcp.Description("The repository name (e.g., 'myorg/myrepo')"),
+		),
+		mcp.WithString("title",
+			mcp.Required(),
+			mcp.Description("The title of the pull request"),
+		),
+		mcp.WithString("body",
+			mcp.Required(),
+			mcp.Description("The body/description of the pull request"),
+		),
+		mcp.WithString("base",
+			mcp.Required(),
+			mcp.Description("The base branch (target branch, usually 'main')"),
+		),
+		mcp.WithString("head",
+			mcp.Required(),
+			mcp.Description("The head branch (source branch with changes)"),
+		),
+	)
+
+	// Add the tool with our handler
+	m.server.AddTool(tool, m.agentPullRequestHandler)
+}
+
+// agentPullRequestHandler - actual implementation with GraphQL call
+func (m *MCPServer) agentPullRequestHandler(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+	// Extract required parameters
+	runId, err := request.RequireString("runId")
```

Comment:
This could be refactored to a tiny function that takes a list of args to validate in a loop.